### ISSUE TITLE
Fixed: During Drupal 7 upgrade switch to the standard installation profile if previous profile is not available.

### ISSUE THREAD

**Issue by SamStamport** [backdrop/backdrop-issues#2518]
When I click the gear icon in the top-right corner of the header block from my home page I get a couple of error messages on-screen and in the log. The error messages also appear when I click admin/reports/status. This began happening once I copied the Files folder from my D7 site into Files in Backdrop and after I replaced the database with the associated D7 database. I guess something went wrong. I hope you can help me sort it out.

I'm using Backdrop 1.6.0 on Windows 10 with XAMPP Version 7.0.13, PHP Version 7.0.13, MariaDB version 10.1.19, Apache version 2.4.23.

Here are the two log messages:

TYPE	php
DATE	Saturday, February 4, 2017 - 12:13pm
USER	admin
LOCATION	http://localhost:8080/BD/home
REFERRER	http://localhost:8080/BD/admin/structure/layouts/manage/home?destination=home
MESSAGE	Notice: Undefined index: name in system_requirements() (line 33 of C:\xampp\htdocs\BD\core\modules\system\system.install).
SEVERITY	notice
HOSTNAME	::1
OPERATIONS	
Home
Content
User accounts
Appearance
Functionality
Structure
Configuration
Reports

Menu search
1
admin
Log out

**and**

TYPE	php
DATE	Saturday, February 4, 2017 - 12:13pm
USER	admin
LOCATION	http://localhost:8080/BD/home
REFERRER	http://localhost:8080/BD/admin/structure/layouts/manage/home?destination=home
MESSAGE	Notice: Undefined index: version in system_requirements() (line 35 of C:\xampp\htdocs\BD\core\modules\system\system.install).
SEVERITY	notice
HOSTNAME	::1
OPERATIONS	

I understand that these messages are classified as severity notice, but they show up in red on in the message area of my site pages indicating something is amiss. Please let me know if you need additional info.

Thanks!

**Comment by jlfranklin** [comment]
On https://localhost:8080/BD/home/admin/reports/status, does it list an "install profile"?  (It would be the second row, right under the Backdrop version.)  What is it?

In the system.core.json file in the active config directory, what is the `install_profile` line?

**Comment by ColumbiaBlooms** [comment]
Install profile is (openoutreach-). This must be debris from my D7 site. It was using the Open Outreach distribution.

In the active system.core.json file I found "install_profile": "openoutreach", There is another system.core.json file in C:\xampp\htdocs\BD\core\modules\system\config\system.core.json with "install_profile": "",

This must be debris from my D7 site. It was using the Open Outreach distribution. When I change the active file from outreach to "" the messages disappear as does the install profile line in the status display. Did I do the right thing?

Why are there two system.core.json files?

**Comment by jlfranklin** [comment]
The one in `system/config` is the one that ships with Backdrop and holds the default values. It is copied into the active config directory during installation. The one in the active config is the live one.

The undefined index errors are because the Open Outreach install profile is not installed.  `system_requirements()` is trying to load the `openoutreach.info` file to get it's human-friendly name and version number.

If leaving it black works, go with it.  The other option is to set it to `standard`.

**Comment by ColumbiaBlooms** [comment]
Thanks so much. 

Problem solved!

**Comment by jlfranklin** [comment]
Glad to hear it!

**Comment by klonos** [comment]
Thanx @jlfranklin for pointing @ColumbiaBlooms to the right direction, but lets keep this issue open and make it so that any missing profiles are handled better and there are more meaningful error messages so that Backdrop doesn't seem "broken". Also tagging as "needs doc update" so to add a step or instructions to our D7 migration guide.

**Comment by ColumbiaBlooms** [comment]
Wow! You guys are awesome. I never got this kind of follow-through with problem posts about Drupal. Please keep up the good work and I'll help you as much as I can.

**Comment by stpaultim** [comment]
I was just doing an experimental upgrade from a D7 site that I quickly built on Pantheon and got this same error message. The site was built using the Pantheon install profile. 

As suggested in this issue, I changed the following setting to: "install_profile": "standard", in the following active config file: system.core.json

Since making this change, I am no longer getting the error message.

**Comment by herbdool** [comment]
Depending on the profile results could be quite different. Here's a couple links that could help. We could port https://www.drupal.org/project/profile_switcher to contrib (originally created to move into a profile but I think works in reverse too). And https://www.davehall.com.au/blog/dave/2012/09/12/switching-installation-profiles-existing-drupal-sites for a couple commands. We'd need to adapt to editing the config file instead of variable table.

**Comment by indigoxela** [comment]
Brainstorming a bit:

In an update hook (D7 to B1 only) we could check for the current profile and if it's missing, switch to "standard". What could be possible side effects?

**Comment by indigoxela** [comment]
A PR is available for testing and review.

It checks for the install profile used in Drupal and if it's not usable in Backdrop (not there, or "minimal") switches it to "standard".

**How to test:**

See [this comment](https://github.com/backdrop/backdrop-issues/issues/4799#issuecomment-743741757) in the other issue for detailed instructions for a testing setup.

TLDR:

The crucial point is: the Drupal install profile is something else than "standard".

Point your Backdrop to the right (Drupal) database, then:

- Set `$settings['update_free_access'] = TRUE;`
- Run /core/update.php (193 pending updates)
- Set `$settings['update_free_access'] = FALSE;`
- Log into your upgraded site
- Go to admin/modules/list - you should *not* see any notices
- Go to admin/reports/status - you should *not* see any notices

Additionally you could inspect your "system" database table:
`SELECT * FROM `system` WHERE `filename` like '%profile';`

Feedback is welcome, testing is extremely appreciated.

**Comment by klonos** [comment]
@indigoxela good timing! ...I'm actually testing this now 🙂 

...I hope you don't mind me editing your previous comment:
- for me it was 197 pending updates
- clarified that the `$settings['update_free_access'] = TRUE;` setting needs to be set in `settings.php`

**Comment by klonos** [comment]
OK, looking at the code, I see that we are basically switching the `minimal` installation profile to `standard`. Although things may work and there may be no errors, we seem to be doing all that in a hacky way (direct db edits). Having a look at all the warnings in https://www.drupal.org/project/profile_switcher makes me worry 🤔 

Taking a step back, isn't this whole thing plain wrong? Shouldn't we be catching this difference, and warning people instead with something like this:

> It seems that you are trying to upgrade a D7 site installed with the minimal profile to a Backdrop site with the standard profile. Things may go wrong when upgrading to mismatching profiles, so it is advised to start with a Backdrop site that is also installed using the minimal profile, and then start the upgrade process over.

Moreover, I think that we should add this as a step to our documentation in https://api.backdropcms.org/documentation/upgrading-from-drupal-7 ...a step to "check your D7 installation profile -> install a matching Backdrop profile if available".

Anyway, having said all the above, I'm not necessarily opposed to this quick-n-dirty profile switching; just skeptical about it, and thinking that we should be doing things as properly as possible. If I'm not mistaken, the `standard` profile = `minimal` profile + additional things. Provided it's not too much work, shouldn't we also be making sure that these additional things that we do as part of the standard profile are also properly installed/enabled during the upgrade/switch? (for example, things like enabling the Project Installer module etc.). Perhaps we need a "proper" minimal to standard conversion here, to ensure that less things might go wrong. Perhaps we should check the code of the https://www.drupal.org/project/profile_switcher module, and incorporate whatever cleanup tasks and failsafe checks make sense into a function that is called from `system_update_1026()`.

...looking at the code of the 7.x version of the module, there's only the .info file + [a single .module file](https://git.drupalcode.org/project/profile_switcher/-/blob/7.x-1.x/profile_switcher.module) with 160 lines of code (including inline comments).

**Comment by klonos** [comment]
...I've compared the PR changes with the code in https://git.drupalcode.org/project/profile_switcher/-/blob/7.x-1.x/profile_switcher.module and see that you are basically doing the same thing @indigoxela 👍 ...I'll test this further and come back with more soon as possible.

**Comment by indigoxela** [comment]
> I've compared the PR changes with the code in https://git.drupalcode.org/project/profile_switcher/-/blob/7.x-1.x/profile_switcher.module and see that you are basically doing the same thing

Right, the profile_switcher module inspired me. It's a little "hacky", that's true, but it's for sure better than forcing people to install the module on their D7 to do the same thing - in an extra step before upgrading.

I suspect, most people aren't aware of the install profile they (or someone else) used to install their Drupal. And they wouldn't know that they have to use profile_switcher in advance. They're probably irritated when they get the notices and have to do some recherche to figure out why.

And we could catch it so easily by updating the entry in the system table and change the config item.

**Comment by klonos** [comment]
> ...but it's for sure better than forcing people to install the module on their D7 to do the same thing - in an extra step before upgrading.

Totally agree 👍 

I'd also add that sites that start with the minimal profile do tend to install more modules as years go by, so they may as well end up being a close match to the standard profile eventually.

**Comment by indigoxela** [comment]
> I'd also add that sites that start with the minimal profile do tend to install more modules as years go by

I did use the minimal profile with Drupal in cases, when I knew that I'll need a totally different setup (no need for page, article, ...)

The main reason for me to also switch from minimal to standard in my PR is, that the minimal profile is a hidden one in Backdrop and it's been discussed to completely remove it at some point. So why worrying people in the future, if we can fix it now.

Again, testing is appreciated.

**Comment by alanmels** [comment]
Please note https://github.com/backdrop-contrib/profile_switcher module had been ported to Backdrop a while ago.

**Comment by indigoxela** [comment]
> profile_switcher module had been ported to Backdrop a while ago

Many thanks for the link. I'm aware of that, but I still think that handling the profile switch in the upgrade process would be a win.

**Comment by alanmels** [comment]
I agree. Posted the link just in case if someone needs it sooner.

**Comment by jenlampton** [comment]
I ran into this today on a Drupal 7 upgrade for a site that started in Minimal. 

Since we have Minimal in Backdrop, that particular issue seems like a bug. Why doesn't minimal return any results for `system_get_info('module', $profile);` (Line 36 of `system.install`)?

```
Notice: Undefined index: name in system_requirements() (line 38 of backdrop/core/modules/system/system.install). 
```

**Comment by indigoxela** [comment]
@jenlampton the "minimal" profile is a hidden one:

```
name = Minimal
description = Start with only a few modules enabled.
version = BACKDROP_VERSION
backdrop = 1.x
type = profile
hidden = TRUE
```
It's not supposed to be selected nor used and is a candidate for getting dropped. That's why my PR switches it to "standard" - to prevent problems.

**Comment by jenlampton** [comment]
All "Hidden" means is that it won't show up in the installer. Minimal is not a candidate for getting dropped, it's there for the experts who would like to be able to use minimal instead of standard. But hiding it from the installer is done so that it doesn't get in the way of novices who won't understand what an install profile is, or why they need to select one just to get Backdrop installed.

Minimal should work as expected.

**Comment by indigoxela** [comment]
> Minimal should work as expected.

Then this should get fixed. Does this make my PR invalid? Not sure.

> Experts will know that they can use...

I'm for sure not a novice, but I didn't know that.

**Comment by indigoxela** [comment]
FTR: this is why I assumed that "minimal" might get removed: https://github.com/backdrop/backdrop-issues/issues/1999

**Comment by jenlampton** [comment]
Thanks, I either hadn't seen that issue, or had forgotten about it!

**Comment by klonos** [comment]
@indigoxela I have just tested your PR because I got the `Notice: Undefined index: name in system_requirements()` error when testing things as part of #5496, and it worked fine 👍🏼 

Setting as WFM, but NW at the same time, as the PR needs a rebase so that tests can run again against the current 1.x and in the newer versions of PHP since 2020 when the initial tests ran (zenci).

**Comment by indigoxela** [comment]
@klonos many thanks for the reminder, PR has been updated. :+1:

**Comment by klonos** [comment]
Thanks @indigoxela 🙏🏼 ...RTBC 👍🏼

**Comment by BWPanda** [comment]
Added a PR review requesting changes.

**Comment by indigoxela** [comment]
@BWPanda many thanks for having a look. :pray:  I'm afraid, I don't currently have the time to evaluate the requested changes, so closing my PR to make space for a better approach.

**Comment by klonos** [comment]
@BWPanda I've created a new PR from @indigoxela's (kept initial commits) + included your feedback/suggestions (with tweaks in the inline comments): https://github.com/backdrop/backdrop/pull/4360

I've re-tested this in the Pantheon scenario, and it still works as expected.

**Comment by BWPanda** [comment]
PR looks good to me. Thanks @indigoxela and @klonos.

**Comment by klonos** [comment]
I'd mark this as WFM as well, but I've worked on the last PR for it, so would be good to have someone else test this as well (perhaps outside the context of a Pantheon hosting).

**Comment by quicksketch** [comment]
Merged https://github.com/backdrop/backdrop/pull/4360 into 1.x and 1.24.x. Thank you so much folks for picking this up again and solving it completely. Thanks @SamStamport for filing this issue.

https://github.com/backdrop/backdrop/commit/269138964f70520e2a7e83cfcd0e04e061edc037 by @indigoxela, @klonos, @BWPanda, @SamStamport, @jlfrankin, @stpaultim, @herbdool, @jenlampton, and @alanmels.
